proposals/0000-scan-takewhile-dropwhile.md

```diff
@@ -0,0 +1,138 @@
+# Feature name
+
+* Proposal: TBD
+* Author(s): [Kevin Ballard](https://github.com/kballard)
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+Add 3 new `SequenceType` functions `scan(_:combine:)`, `takeWhile(_:)`, and
+`dropWhile(_:)`, with overrides as appropriate on `CollectionType`,
+`LazySequenceType`, and `LazyCollectionType`, as well as a global function
+`iterate(_:apply:)`.
+
+Swift-evolution thread:
+[Proposal: Add scan, takeWhile, dropWhile, and iterate to the stdlib](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20151228/004690.html)
+
+## Motivation
+
+The Swift standard library provides many useful sequence manipulators like
+`dropFirst(_:)`, `filter(_:)`, etc. but it's missing a few common methods that
+are quite useful.
+
+## Proposed solution
+
+Add the following extension to `SequenceType`:
+
+```swift
+extension SequenceType {
+  /// Returns an array containing the results of
+  ///
+  ///     p.reduce(initial, combine: combine)
+  ///
+  /// for each prefix `p` of `self`, in order from shortest to longest.
+  /// For example:
+  ///
+  ///     (1..<6).scan(0, combine: +) // [0, 1, 3, 6, 10, 15]
+  ///
+  /// - Complexity: O(N)
+  func scan<T>(initial: T, @noescape combine: (T, Self.Generator.Element) throws -> T) rethrows -> [T]
+}
+```
+
+Modify the declaration of `SequenceType` with two new members:
+
+```swift
+protocol SequenceType {
+  // ...
+  /// Returns a subsequence by skipping elements while `dropElement` returns
+  /// `true` and returning the remainder.
```

Comment:
I don't like the emphasis on suffixing here. When you think "suffix" you typically think about working backwards from the end, but this function is really about skipping elements at the start. And in fact the name `suffixFrom(firstElementSatisfying:)` could be construed as starting with the first element from the end that satisfies that predicate, instead of the first element from the start.

As for the bit about returns a subsequence, you may as well say that `map(_:)` shouldn't say "Return an `Array` ..." since that's part of the type signature.